taichi/program/sparse_solver.cpp

```diff
@@ -55,20 +57,23 @@ bool EigenSparseSolver<EigenSolver>::info() {
   return solver_.info() == Eigen::Success;
 }
 
-std::unique_ptr<SparseSolver> make_sparse_solver(const std::string &solver_type,
+std::unique_ptr<SparseSolver> make_sparse_solver(DataType dt,
+                                                 const std::string &solver_type,
                                                  const std::string &ordering) {
-  using key_type = std::pair<std::string, std::string>;
+  using key_type = Triplets;
   using func_type = std::unique_ptr<SparseSolver> (*)();
-  static const std::unordered_map<key_type, func_type, pair_hash>
+  static const std::unordered_map<key_type, func_type, key_hash>
       solver_factory = {
-          MAKE_SOLVER(LLT, AMD),
-          MAKE_SOLVER(LLT, COLAMD),
-          MAKE_SOLVER(LDLT, AMD),
-          MAKE_SOLVER(LDLT, COLAMD),
-      };
-  std::pair<std::string, std::string> solver_key =
-      std::make_pair(solver_type, ordering);
+          MAKE_SOLVER(float32, LLT, AMD), MAKE_SOLVER(float32, LLT, COLAMD),
+          MAKE_SOLVER(float32, LDLT, AMD), MAKE_SOLVER(float32, LDLT, COLAMD)};
+  static const std::unordered_map<std::string, std::string> dt_map = {
+      {"f32", "float32"}, {"f64", "float64"}};
```

Comment:
Is float64 type supported now? Should we add float64 in `solver_factory`? (or maybe later after float64 is fully supported.